### ISSUE TITLE
Expose variables and parameters of a problem using a dict

### ISSUE THREAD

**Issue by tschm** [cvxgrp/cvxpy#966]
**Is your feature request related to a problem? Please describe.**
A clear and concise description of what the problem is. 

Currently variables and parameters are exposed as lists of objects. This is somewhat clumsy. Here's a test problem:
```
import cvxpy as cvx
import numpy as np

from pyneutral.util import Solver


def solver(n):
    weights = cvx.Variable(n, name="weights")

    # The return will be equal to the weights times the expected return mu, for each asset
    mu = cvx.Parameter(n, name="exp_return", value=np.ones(n))
    cov_matrix = cvx.Parameter(shape=(n, n), name="cov_matrix", PSD=True, value=np.identity(n))
    max_std = cvx.Parameter(nonneg=True, name="max_std", value=0.0)

    # the objective
    ret = mu * weights

    constraints = [cvx.sum(weights) == 0,
                   cvx.quad_form(weights, cov_matrix) <= max_std]

    return Solver(problem=cvx.Problem(objective=cvx.Maximize(ret), constraints=constraints))
```

**Describe the solution you'd like**
A clear and concise description of what you want to happen, and how your solution would solve the problem.

Note that the problem is hidden in a function and changing the parameters (there are 3 of them) is complicated. Hence I wrote the Solver object

```
class Solver(object):
    def __init__(self, problem):
        assert isinstance(problem, cvx.Problem)
        self.problem = problem
        self.parameters = {a.name(): a for a in problem.parameters()}
        self.variables = {a.name(): a for a in problem.variables()}

    def solve(self):
        self.problem.solve()
        # Check that the solver computed an optimal solution
        assert self.problem.status == "optimal", "Couldn't find an optimal solution. Status {s}".format(
            s=self.problem.status)
```

**Additional context**
It's now easy to modify the parameters on the fly

```
if __name__ == '__main__':
    # only construct the problem
    problem = solver(2)

    # Inject values for the parameters used in the problem constructed above
    problem.parameters["exp_return"].value = np.array([2.0, 3.0])
    problem.parameters["max_std"].value = 20.0
    problem.parameters["cov_matrix"].value = np.array([[1.0, 0.0], [0.0, 1.0]])

    # Only now we solve the problem
    problem.solve()

    print(problem.variables["weights"].value)
```

Do I miss something here?

**Comment by akshayka** [comment]
We could perhaps add methods like `get_variable` and `get_parameter`, which return a variable/parameter by its name. 

I might have a slight preference to encourage users to keep track of their variables and parameters explicitly. If you need to track the names, might as well track the objects themselves. But I’d be open to adding these two methods.

**Comment by bstellato** [comment]
> We could perhaps add methods like get_variable and get_parameter, which return a variable/parameter by its name.

I like this idea!


I think it would also be great to automatically assign the attribute `name` to the actual Python variable name. However, it might be clumsy to do in Python (in Julia for example, it can be done with a macro).

**Comment by akshayka** [comment]
> I think it would also be great to automatically assign the attribute name to the actual Python variable name.

@bstellato , can you give an example of what you mean by this?

**Comment by tschm** [comment]
I am not a fan of get_variable or get_parameter. Looks like Java to me and in particular you don't know what names are available without using other methods to find out. The dictionary would give you such luxury for free. Since you have already parameters() returning a list you may want to do parameters_dict() and parameters_list(). I would give parameters() a deprecation warning.

**Comment by akshayka** [comment]
I agree with you that get_variable is clunky; it’s also reminiscent of TF 1.0, which is not necessarily a good thing ...

I’m open to switching parameters/variables to dicts sometime in the future. That’d be a very breaking change, something for CVXPY 2.0. I guess in the meantime we could do something similar to what you suggest.

**Comment by tschm** [comment]
I would just introduce the properties "parameter" and "variable" both returning dictionaries. Then you could keep the parameters and variables methods and the change wouldn't be breaking at all.

**Comment by rileyjmurray** [comment]
> > I think it would also be great to automatically assign the attribute name to the actual Python variable name.
> 
> @bstellato , can you give an example of what you mean by this?

I think he means that if a user calls ``x = cp.Variable()``, then we automatically set ``x._name = "x"``. Personally I don't like such an approach, because it means we can't rely on variable names serving as unique identifiers in a model.

**Comment by rileyjmurray** [comment]
> I would just introduce the properties "parameter" and "variable" both returning dictionaries. Then you could keep the "parameters" and "variables" methods and the change wouldn't be breaking at all.

@tschm I like this. If they look too similar, the names of these dict-valued properties could be shortened further still (for example, ``param`` and ``var``) or appended with a suffix (e.g. ``par_dict``, ``var_dict``, perhaps without underscores).

**Comment by tschm** [comment]
I guess the dictionary approach won't work if people don't assign names to their variables and parameters. Throw an error or less attractive: we shouldn't make the name an optional argument in the constructor for variables and parameters? Let me at this stage also mention that I am doing convex programming ever since 2007. It's such a pleasure to do it today with cvxpy. It's incredible powerful, readable and easy to use. Great work guys.

**Comment by SteveDiamond** [comment]
Thanks @tschm ! Variables and parameters are assigned names automatically if the user doesn't provide one. Internally we use dictionaries of variable/parameter id to object a lot. The distinction was created because in theory someone could give two different variables the same name.

**Comment by luismiguel-lopez** [comment]
Hi, I agree it would be really useful to have a method to quickly return all variables and/or parameters in a dictionary. Are you planning to add it? Thanks.

**Comment by SteveDiamond** [comment]
Yeah it looks like we're all on board. Maybe @tschm could submit a PR?

**Comment by tschm** [comment]
I am very sorry. I don't think I should add a little PR. I am happy to share a larger code fragment and leave it to the experts to rip it into parts:

```
from dataclasses import dataclass
from typing import List
import cvxpy as cvx


@dataclass(init=False)
class Builder:
    objective: cvx.problems.objective.Objective = None
    constraints: List[cvx.constraints.constraint.Constraint] = None

    def build(self):
        assert self.objective is not None, "Please define a valid objective."
        assert self.constraints is not None, \
            "Please set explicitly self.constraints = [] if you intend to work without " \
            "constraints. Otherwise please check the spelling!"

        return Problem(problem=cvx.Problem(objective=self.objective, constraints=self.constraints))


@dataclass(frozen=True)
class Problem:
    problem: cvx.Problem

    def solve(self, *args, **kwargs):
        self.problem.solve(*args, **kwargs)
        # Check that the solver computed an optimal solution
        assert self.problem.status == "optimal", \
            "Couldn't find an optimal solution. Status {s}". \
                format(s=self.problem.status)

    @property
    def parameters(self):
        return {a.name(): a for a in self.problem.parameters()}

    @property
    def variables(self):
        return {a.name(): a for a in self.problem.variables()}

    def __repr__(self):
        return str(self.problem)
```
Here's an example:

```
import cvxpy as cvx
import numpy as np
import pytest

from pyconvex.convex.problem import Builder


class Markowitz(Builder):
    def __init__(self, n):
        self.weights = cvx.Variable(shape=n, name="weights")
        self.mu = cvx.Parameter(shape=n, name="expected return")
        self.cov = cvx.Parameter(shape=(n, n), name="covariance", PSD=True, value=np.identity(n))

    @property
    def funding(self):
        return cvx.sum(self.weights)

    @property
    def leverage(self):
        return self.norm(p=1)

    @property
    def variance(self):
        return cvx.quad_form(self.weights, self.cov)

    @property
    def expected_return(self):
        return self.mu @ self.weights

    def norm(self, p):
        return cvx.norm(self.weights, p=p)


def test_markowitz_1():
    m = Markowitz(n=2)
    m.mu.value = np.array([1.0, 3.0])
    m.objective = cvx.Maximize(m.expected_return)
    m.constraints = [m.funding == 1, m.weights >= 0]

    problem = m.build()
    problem.solve()
    np.testing.assert_almost_equal(problem.variables["weights"].value, np.array([0.0, 1.0]))
    np.testing.assert_almost_equal(m.weights.value, np.array([0.0, 1.0]))
```

**Comment by tschm** [comment]
What I like about the approach above is that one can really separate the definition of the problem from the injection of data. I am using dataclass but I don't have to accomodate users with somewhat old versions of Python. I assume you don't want to have that in a library as popular as cvxpy.

**Comment by tschm** [comment]
To reinforce my point about separating definition and injection one can write also

```
def test_markowitz_1():
    m = Markowitz(n=2)
    m.objective = cvx.Maximize(m.expected_return)
    m.constraints = [m.funding == 1, m.weights >= 0]
    problem = m.build()

    problem.parameter["expected return"].value = np.array([1.0, 3.0])
    problem.solve()

    np.testing.assert_almost_equal(problem.variables["weights"].value, np.array([0.0, 1.0]))
    np.testing.assert_almost_equal(m.weights.value, np.array([0.0, 1.0]))
```

**Comment by SteveDiamond** [comment]
Ok thanks @tschm certainly no need to do a PR if you don't want to. This is more than enough for us to work with :)

**Comment by tschm** [comment]
Sorry, it's more a skill rather than a motivational thing :-) But I have done it now and this is my first PR on such a prominent package. I hope you can make something with it. I should have added a test though. Can do in a 2nd PR

**Comment by SteveDiamond** [comment]
Closed by #1279
